src/sbin/tunefs/tunefs.8
.\"	$OpenBSD: tunefs.8,v 1.22 2004/07/28 21:27:52 jmc Exp $
.\"	$NetBSD: tunefs.8,v 1.8 1995/03/18 15:01:29 cgd Exp $
.\"
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)tunefs.8	8.2 (Berkeley) 12/11/93
.\"
.Dd December 11, 1993
.Dt TUNEFS 8
.Os
.Sh NAME
.Nm tunefs
.Nd tune up an existing filesystem
.Sh SYNOPSIS
.Nm tunefs
.Op Fl Ap
.Op Fl a Ar maxcontig
.Op Fl d Ar rotdelay
.Op Fl e Ar maxbpg
.Op Fl f Ar avgfilesize
.Bk -words
.Op Fl m Ar minfree
.Ek
.Op Fl n Ar avgfpdir
.Bk -words
.Op Fl o Ar optimize_preference
.Ek
.Bk -words
.Op Ar special | Ar filesys
.Ek
.Sh DESCRIPTION
.Nm
is designed to change the dynamic parameters of a filesystem
which affect the layout policies.
The parameters which are to be changed are indicated by the flags
given below:
.Bl -tag -width Ds
.It Fl A
The filesystem has several backups of the super-block.
Specifying this option will cause all backups to be modified as well as the
primary super-block.
This is potentially dangerous - use with caution.
.It Fl a Ar maxcontig
This specifies the maximum number of contiguous blocks that will
be laid out before forcing a rotational delay (see
.Fl d
below).
The default value depends on the block size
(4 for 16KB blocks, 8 for 8KB blocks and 16 for 4KB blocks).
Most device drivers can chain several buffers together in a single transfer.
For optimal performance, the value of
.Ar maxcontig
should be chosen based on the maximum chain length supported by
the device driver.
.It Fl d Ar rotdelay
This specifies the expected time (in milliseconds)
to service a transfer completion
interrupt and initiate a new transfer on the same disk.
It is used to decide how much rotational spacing to place between
successive blocks in a file.
.It Fl e Ar maxbpg
This indicates the maximum number of blocks any single file can
allocate out of a cylinder group before it is forced to begin
allocating blocks from another cylinder group.
Typically this value is set to about one quarter of the total blocks
in a cylinder group.
The intent is to prevent any single file from using up all the
blocks in a single cylinder group,
thus degrading access times for all files subsequently allocated
in that cylinder group.
The effect of this limit is to cause big files to do long seeks
more frequently than if they were allowed to allocate all the blocks
in a cylinder group before seeking elsewhere.
For filesystems with exclusively large files,
this parameter should be set higher.
.It Fl f Ar avgfilesize
Specifies the expected average file size in bytes.
This value could be used for various optimizations, but for now it is only
used together with
.Ar avgfpdir
to optimize the directory allocation policy.
To take effect, both
.Ar avgfpdir
and
.Ar avgfilesize
must be greater than zero.
(Also see
.Ar avgfpdir . )
.It Fl m Ar minfree
This value specifies the percentage of space held back
from normal users; the minimum free space threshold.
The default value used is 5%.
This value can be set to zero; however, a factor of up to three
in throughput will be lost over the performance obtained at a 5%
threshold.
Note that if the value is raised above the current usage level,
users will be unable to allocate files until enough files have
been deleted to get under the higher threshold.
.It Fl n Ar avgfpdir
Specifies the expected average number of files per directory in
the filesystem.
This value is used only if both
.Ar avgfilesize
and
.Ar avgfpdir
are greater than zero.
It is used to limit number of directories which may be allocated one after
another in the same cylinder group without intervention by regular files.
This value does not affect most filesystem operations but is useful for
applications which at first create a directory structure and then populate
with files later.
(Also see
.Ar avgfilesize . )
.It Fl o Ar optimize_preference
The filesystem can either try to minimize the time spent
allocating blocks, or it can attempt to minimize the space
fragmentation on the disk.
If the value of
.Ar minfree
(see above) is less than 5%,
then the filesystem should optimize for space to avoid
running out of full sized blocks.
For values of
.Ar minfree
greater than or equal to 5%,
fragmentation is unlikely to be problematical, and
the filesystem can be optimized for time.
.It Fl p
This option shows a summary of what the current tuneable settings
are on the selected filesystem.
More detailed information can be obtained in the
.Xr dumpfs 8
manual page.
.El
.Sh SEE ALSO
.Xr fs 5 ,
.Xr dumpfs 8 ,
.Xr growfs 8 ,
.Xr newfs 8
.Rs
.%A M. McKusick
.%A W. Joy
.%A S. Leffler
.%A R. Fabry
.%T "A Fast Filesystem for UNIX"
.%J "ACM Transactions on Computer Systems 2"
.%N 3
.%P pp 181-197
.%D August 1984
.%O "(reprinted in the BSD System Manager's Manual, SMM:5)"
.Re
.Sh HISTORY
The
.Nm
command appeared in
.Bx 4.2 .
.Sh BUGS
This program should work on mounted and active filesystems.
Because the super-block is not kept in the buffer cache,
the changes will only take effect if the program
is run on dismounted filesystems.
To change the root filesystem, the system must be rebooted
after the filesystem is tuned.
.Pp
.\" Take this out and a Unix Demon will dog your steps from now until
.\" the time_t's wrap around.
You can tune a filesystem, but you can't tune a fish.


src/sbin/tunefs/tunefs.c
/*	$OpenBSD: tunefs.c,v 1.23 2003/07/10 00:00:58 david Exp $	*/
/*	$NetBSD: tunefs.c,v 1.10 1995/03/18 15:01:31 cgd Exp $	*/

/*
 * Copyright (c) 1983, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1983, 1993\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)tunefs.c	8.2 (Berkeley) 4/19/94";
#else
static const char rcsid[] =
	"$OpenBSD: tunefs.c,v 1.23 2003/07/10 00:00:58 david Exp $";
#endif
#endif /* not lint */

/*
 * tunefs: change layout parameters to an existing filesystem.
 */
#include <sys/param.h>
#include <sys/stat.h>

#include <ufs/ffs/fs.h>

#include <errno.h>
#include <err.h>
#include <fcntl.h>
#include <fstab.h>
#include <stdio.h>
#include <paths.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

/* the optimization warning string template */
#define	OPTWARN	"should optimize for %s with minfree %s %d%%"

union {
	struct	fs sb;
	char pad[MAXBSIZE];
} sbun;
#define	sblock sbun.sb

int fi = -1;
long dev_bsize = 1;

void bwrite(daddr_t, char *, int);
int bread(daddr_t, char *, int);
void getsb(struct fs *, char *, int);
void usage(void);
void printfs(void);

extern char *__progname;

int
main(int argc, char *argv[])
{
	char *cp, *special, *name;
	struct stat st;
	int i;
	int Aflag = 0;
	struct fstab *fs;
	char *chg[2], device[MAXPATHLEN];

	argc--, argv++;
	if (argc < 2)
		usage();
	special = argv[argc - 1];
	fs = getfsfile(special);
	if (fs)
		special = fs->fs_spec;
again:
	if (stat(special, &st) < 0) {
		if (*special != '/') {
			if (*special == 'r')
				special++;
			(void)snprintf(device, sizeof(device), "%s%s",
				       _PATH_DEV, special);
			special = device;
			goto again;
		}
		err(1, "%s", special);
	}
	if (!S_ISBLK(st.st_mode) && !S_ISCHR(st.st_mode))
		errx(10, "%s: not a block or character device", special);
	for (; argc > 0 && argv[0][0] == '-'; argc--, argv++) {
		for (cp = &argv[0][1]; *cp; cp++)
			switch (*cp) {

			case 'A':
				Aflag++;
				continue;

			case 'p':
				getsb(&sblock, special, O_RDONLY);
				printfs();
				return (0);

			case 'a':
				getsb(&sblock, special, O_RDWR);
				name = "maximum contiguous block count";
				if (argc < 1)
					errx(10, "-a: missing %s", name);
				argc--, argv++;
				i = atoi(*argv);
				if (i < 1)
					errx(10, "%s must be >= 1 (was %s)",
					    name, *argv);
				warnx("%s changes from %d to %d",
				    name, sblock.fs_maxcontig, i);
				sblock.fs_maxcontig = i;
				continue;

			case 'd':
				getsb(&sblock, special, O_RDWR);
				name =
				   "rotational delay between contiguous blocks";
				if (argc < 1)
					errx(10, "-d: missing %s", name);
				argc--, argv++;
				i = atoi(*argv);
				warnx("%s changes from %dms to %dms",
				    name, sblock.fs_rotdelay, i);
				sblock.fs_rotdelay = i;
				continue;

			case 'e':
				getsb(&sblock, special, O_RDWR);
				name =
				  "maximum blocks per file in a cylinder group";
				if (argc < 1)
					errx(10, "-e: missing %s", name);
				argc--, argv++;
				i = atoi(*argv);
				if (i < 1)
					errx(10, "%s must be >= 1 (was %s)",
					    name, *argv);
				warnx("%s changes from %d to %d",
				    name, sblock.fs_maxbpg, i);
				sblock.fs_maxbpg = i;
				continue;

			case 'f':
				getsb(&sblock, special, O_RDWR);
				name = "average file size";
				if (argc < 1)
					errx(10, "-f: missing %s", name);
				argc--, argv++;
				i = atoi(*argv);
				if (i < 0)
					errx(10, "%s must be >= 0 (was %s)",
					    name, *argv);
				warnx("%s changes from %d to %d",
				    name, sblock.fs_avgfilesize, i);
				sblock.fs_avgfilesize = i;
				continue;

			case 'm':
				getsb(&sblock, special, O_RDWR);
				name = "minimum percentage of free space";
				if (argc < 1)
					errx(10, "-m: missing %s", name);
				argc--, argv++;
				i = atoi(*argv);
				if (i < 0 || i > 99)
					errx(10, "bad %s (%s)", name, *argv);
				warnx("%s changes from %d%% to %d%%",
				    name, sblock.fs_minfree, i);
				sblock.fs_minfree = i;
				if (i >= MINFREE &&
				    sblock.fs_optim == FS_OPTSPACE)
					warnx(OPTWARN, "time", ">=", MINFREE);
				if (i < MINFREE &&
				    sblock.fs_optim == FS_OPTTIME)
					warnx(OPTWARN, "space", "<", MINFREE);
				continue;

			case 'n':
				getsb(&sblock, special, O_RDWR);
				name = "expected number of files per directory";
				if (argc < 1)
					errx(10, "-n: missing %s", name);
				argc--, argv++;
				i = atoi(*argv);
				if (i < 0)
					errx(10, "%s must be >= 0 (was %s)",
					    name, *argv);
				warnx("%s changes from %d to %d",
				    name, sblock.fs_avgfpdir, i);
				sblock.fs_avgfpdir = i;
				continue;

			case 's':
				errx(1, "See mount(8) for details about"
				      " how to enable soft updates.");

			case 'o':
				getsb(&sblock, special, O_RDWR);
				name = "optimization preference";
				if (argc < 1)
					errx(10, "-o: missing %s", name);
				argc--, argv++;
				chg[FS_OPTSPACE] = "space";
				chg[FS_OPTTIME] = "time";
				if (strcmp(*argv, chg[FS_OPTSPACE]) == 0)
					i = FS_OPTSPACE;
				else if (strcmp(*argv, chg[FS_OPTTIME]) == 0)
					i = FS_OPTTIME;
				else
					errx(10, "bad %s (options are `space' or `time')",
					    name);
				if (sblock.fs_optim == i) {
					warnx("%s remains unchanged as %s",
					    name, chg[i]);
					continue;
				}
				warnx("%s changes from %s to %s",
				    name, chg[sblock.fs_optim], chg[i]);
				sblock.fs_optim = i;
				if (sblock.fs_minfree >= MINFREE &&
				    i == FS_OPTSPACE)
					warnx(OPTWARN, "time", ">=", MINFREE);
				if (sblock.fs_minfree < MINFREE &&
				    i == FS_OPTTIME)
					warnx(OPTWARN, "space", "<", MINFREE);
				continue;

			default:
				usage();
			}
	}
	if (argc != 1)
		usage();
	bwrite((daddr_t)SBOFF / dev_bsize, (char *)&sblock, SBSIZE);
	if (Aflag)
		for (i = 0; i < sblock.fs_ncg; i++)
			bwrite(fsbtodb(&sblock, cgsblock(&sblock, i)),
			    (char *)&sblock, SBSIZE);
	if (close(fi))
		err(1, "close: %s", special);

	return (0);
}

void
usage(void)
{
	fprintf(stderr,
		"Usage: %s tuneup-options special-device\n"
		"where tuneup-options are:\n"
		"\t-A modify all backups of the super-block\n"
		"\t-a maximum contiguous blocks\n"
		"\t-d rotational delay between contiguous blocks\n"
		"\t-e maximum blocks per file in a cylinder group\n"
		"\t-f expected average file size\n"
		"\t-m minimum percentage of free space\n"
		"\t-n expected number of files per directory\n"
		"\t-o optimization preference (`space' or `time')\n"
		"\t-p no change - just prints current tuneable settings\n",
		__progname);
	exit(2);
}

void
getsb(struct fs *fs, char *file, int flags)
{

	if (fi >= 0)
		return;
	fi = open(file, flags);
	if (fi < 0)
		err(3, "cannot open %s", file);
	if (bread((daddr_t)SBOFF, (char *)fs, SBSIZE))
		err(4, "%s: bad super block", file);
	if (fs->fs_magic != FS_MAGIC)
		errx(5, "%s: bad magic number", file);
	dev_bsize = fs->fs_fsize / fsbtodb(fs, 1);
}

void
printfs(void)
{
	warnx("maximum contiguous block count: (-a)               %d",
	      sblock.fs_maxcontig);
	warnx("rotational delay between contiguous blocks: (-d)   %d ms",
	      sblock.fs_rotdelay);
	warnx("maximum blocks per file in a cylinder group: (-e)  %d",
	      sblock.fs_maxbpg);
	warnx("expected average file size: (-f)                   %d",
	      sblock.fs_avgfilesize);
	warnx("minimum percentage of free space: (-m)             %d%%",
	      sblock.fs_minfree);
	warnx("expected number of files per directory: (-n)       %d",
	      sblock.fs_avgfpdir);
	warnx("optimization preference: (-o)                      %s",
	      sblock.fs_optim == FS_OPTSPACE ? "space" : "time");
	if (sblock.fs_minfree >= MINFREE &&
	    sblock.fs_optim == FS_OPTSPACE)
		warnx(OPTWARN, "time", ">=", MINFREE);
	if (sblock.fs_minfree < MINFREE &&
	    sblock.fs_optim == FS_OPTTIME)
		warnx(OPTWARN, "space", "<", MINFREE);
}

void
bwrite(daddr_t blk, char *buf, int size)
{

	if (lseek(fi, (off_t)blk * dev_bsize, SEEK_SET) < 0)
		err(6, "FS SEEK");
	if (write(fi, buf, size) != size)
		err(7, "FS WRITE");
}

int
bread(daddr_t bno, char *buf, int cnt)
{
	int i;

	if (lseek(fi, (off_t)bno * dev_bsize, SEEK_SET) < 0)
		return(1);
	if ((i = read(fi, buf, cnt)) != cnt) {
		for(i=0; i<sblock.fs_bsize; i++)
			buf[i] = 0;
		return (1);
	}
	return (0);
}
